### PR TITLE
Fully codable model

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
I've basically been going through the next chunk of code to document which is the Codable implementation of the model. I've added documentation, done some refactoring, and added some fully automated tests to try and catch any issues that are introduced during refactoring. (Manually created tests tend to only catch those things that I am expecting.)

Anyhow, it's a lot of changes. (sorry) I've been trying to work towards being more consistent and putting in support for generating documentation of the json schema. (work in progress)

I have more to do but this is a lot of changes so thought I'd submit. 

Oh, and I am making changes here that will break CRF but will fix in CRF once this is in master. Right now, I am going with the rational that the more I can get cleaned up before we release to outside developers, the less we have to worry about deprecation warnings later.

### CODE REVIEW

File: ResearchSuite/ResearchSuite/RSDFactory.swift [hunk 1]
```diff
@@ -114,6 +114,35 @@ open class RSDFactory {
...
+    /// Decode the schema info from this decoder. This method *must* return a schema info object.
+    /// The default implementation will return a `RSDSchemaInfoObject`.
+    /// - parameter decoder: The decoder to use to instatiate the object.
+    /// - returns: The schema info created from this decoder.
+    /// - throws: `DecodingError` if the object cannot be decoded.
+    open func decodeSchemaInfo(from decoder: Decoder) throws -> RSDSchemaInfo {
+        return try RSDSchemaInfoObject(from: decoder)
+    }
+    
+    /// Encode the schma info from the given task result to the given encoder. This allows a subclass
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDFactory.swift]: schma -> schema

File: ResearchSuite/ResearchSuite/RSDFactory.swift [hunk 2]
```diff
@@ -504,10 +498,53 @@ open class RSDFactory {
...
+        if hasDateComponents && hasTimeComponents {
+            return timestampFormatter
+        } else if hasTimeComponents {
+            return timeOnlyFormatter
+        } else {
+            return dateOnlyFormatter
+        }
+    }
+    
+    /// `DateFormatter` to use for coding data-only strings. Default = `rsd_ISO8601DateOnlyFormatter`.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDFactory.swift]: data-only --> date-only

File: ResearchSuite/ResearchSuite/RSDInputField.swift [hunk 3]
```diff
@@ -83,8 +85,9 @@ public protocol RSDInputField {
 /// `RSDChoice` is used to describe a choice item for use with a multiple choice or multiple component input field.
 public protocol RSDChoice {
     
-    /// A JSON encodable object to return as the value when this choice is selected.
-    var value: Codable { get }
+    /// A JSON encodable object to return as the value when this choice is selected. A `nil` value indicates that
+    /// the user has selected to skip the question or "prefers not to answer".
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDInputField.swift:89]: I thought "prefers not to answer" was usually implemented as a valid choice with its own value, to distinguish from just skipping the question altogether...?

File: ResearchSuite/ResearchSuite/RSDInputField.swift [hunk 4]
```diff
@@ -104,6 +107,11 @@ public protocol RSDChoice {
     ///     - size:        The size of the image to return.
     ///     - callback:    The callback with the image, run on the main thread.
     func fetchIcon(for size: CGSize, callback: @escaping ((UIImage?) -> Void))
+    
+    /// Is the choice value equal to the given result?
+    /// - parameter result: An result to test for equality.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDInputField.swift]: An --> A

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 5]
```diff
@@ -132,31 +145,66 @@ public struct RSDMeasurementPickerDataSourceObject : RSDChoicePickerDataSource {
...
+    /// The type of UI picker to display for dates and times.
     public let datePickerMode: RSDDatePickerMode
+
+    /// Specify the minimum date range. Default = `nil`. When `minimumDate` > `maximumDate`, the values are ignored.
     public let minimumDate: Date?
+
+    /// Specify the maximum date range. Default = `nil`. When `minimumDate` > `maximumDate`, the values are ignored.
     public let maximumDate: Date?
+
+    /// display minutes wheel with interval. interval must be evenly divided into 60. default is 1. min is 1, max is 30
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: Capitalization

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 6]
```diff
@@ -132,31 +145,66 @@ public struct RSDMeasurementPickerDataSourceObject : RSDChoicePickerDataSource {
...
+        self.dateFormatter = dateFormatter
+    }
 }
 
+/// A simple struct that can be used to implement the `RSDDateComponentPickerDataSource` protocol.
 public struct RSDDateComponentPickerDataSourceObject : RSDDateComponentPickerDataSource {
+    /// The calendar to use for the date components
     public let calendar: Calendar
+
+    /// The components to include in the picker
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: .

File: ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift [hunk 7]
```diff
@@ -132,31 +145,66 @@ public struct RSDMeasurementPickerDataSourceObject : RSDChoicePickerDataSource {
...
     public let calendarComponents: Set<Calendar.Component>
+
+    /// The minimum year if the year is included, otherwise this value is ignored.
     public let minimumYear: Int?
+
+    /// The maximum year if the year is included, otherwise this value is ignored.
     public let maximumYear: Int?
+
+    /// The date components formatter for displaying the date components in a text field or label
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDPickerDataSourceObject.swift]: .

File: ResearchSuite/ResearchSuite/RSDRangeObject.swift [hunk 8]
```diff
@@ -105,42 +140,137 @@ public struct RSDDateRangeObject : RSDDateRange, Codable {
...
         self.minimumValue = minimumDecimal
         self.maximumValue = maximumDecimal
         self.stepInterval = stepInterval
         self.unit = unit
-        self.formatter = numberFormatter
+        self.formatter = formatter
     }
     
-    public init(minimumInt: Int?, maximumInt: Int?, stepInterval: Int? = nil, unit: String? = nil, numberFormatter: NumberFormatter? = nil) {
+    /// Default initializer for an `Int` range. This is used to initialize the range for a `Int` type.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDRangeObject.swift]: a --> an

File: ResearchSuite/ResearchSuite/RSDStandardPermissionType.swift [hunk 9]
```diff
@@ -33,28 +33,21 @@
...
-}
-
-/// Standard permission types
+/// Standard permission types.
+///
+/// - note: This framework intentionally does not include any direct reference to Health Kit.
+///         First, including Health Kit in applications that do not use that SDK makes it
+///         confusing and difficult for researchers to set up the app. Second, the goal of
+///         this framework is to include a model that is platform-agnostic and can be used
+///         independantly of the device. (syoung 11/1/7/2017)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDStandardPermissionType.swift]: independ**e**ntly

File: ResearchSuite/ResearchSuite/RSDStepType.swift [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    /// Defaults to creating a `RSDActiveUIStep` used as a countdown to an active step.
+    public static let countdown: RSDStepType = "countdown"
+    
+    /// Defaults to creating a `RSDFormUIStep`.
+    public static let form: RSDStepType = "form"
+    
+    /// Defaults to creating a `RSDUIStep`.
+    public static let instruction: RSDStepType = "instruction"
+    
+    /// Defaults to creating a `RSDSectionStep`
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDStepType.swift]: .

File: ResearchSuite/ResearchSuite/RSDStepType.swift [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    public static let section: RSDStepType = "section"
+    
+    /// Defaults to creating a `RSDSectionStep` created using a `RSDTransformerStep`
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDStepType.swift]: .

File: ResearchSuite/ResearchSuite/RSDStepType.swift [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    public static let transform: RSDStepType = "transform"
+    
+    /// Defaults to creating a `RSDTaskInfoStep`.
+    public static let taskInfo: RSDStepType = "taskInfo"
+    
+    /// List of all the standard types
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDStepType.swift]: .

File: ResearchSuite/ResearchSuite/RSDSurveyRule.swift [hunk 13]
```diff
@@ -106,25 +114,41 @@ extension RSDComparableSurveyRule {
...
         let skipTo: String = skipToIdentifier ?? RSDIdentifier.exit.rawValue
         
+        return isMatching(to: result, op: op) ? skipTo : nil
+    }
+}
+
+extension RSDComparable {
+    
+    /// Is the choice value equal to the given result?
+    /// - parameter result: An result to test for equality.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDSurveyRule.swift]: An --> A

File: ResearchSuite/ResearchSuite/RSDTableItem.swift [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    
+    /// The choice for a single or multiple choice input field.
+    open private(set) var choice: RSDChoice
+    
+    /// Whether or not the choice is currently selected.
+    public var selected: Bool = false
+    
+    /// Initialize a new RSDChoiceTableItem.
+    /// parameters:
+    ///     - rowIndex:      The index of this item relative to all rows in the section in which this item resides
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItem.swift]: .

File: ResearchSuite/ResearchSuite/RSDTableItemGroup.swift [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+    /// set the answer.
+    /// - parameter newValue: The new value for the answer.
+    /// - throws: `RSDInputFieldError` if the answer is invalid.
+    public final func setAnswer(_ newValue: Any?) throws {
+        _answer = try validatedAnswer(newValue)
+    }
+    
+    /// Set the new answer value from a previous result. This will throw an error if the result isn't valid.
+    /// Otherwise, it will set the answer.
+    /// - parameter result: The result that *may have a previous answer.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItemGroup.swift]: *may ?

File: ResearchSuite/ResearchSuite/RSDTableItemGroup.swift [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+            else {
+                let context = RSDInputFieldError.Context(identifier: inputField.identifier, value: answer, answerResult: answerType, debugDescription: "Result answer type for \(result) not expected type.")
+                throw RSDInputFieldError.invalidType(context)
+        }
+        try self.setAnswer(answerResult.value)
+    }
+    
+    /// Determine if the current answer is valid. Also checks the case where answer is required but one has
+    /// not been provided.
+    /// - returns: A Bool indicating if answer is valid
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItemGroup.swift]: .

File: ResearchSuite/ResearchSuite/RSDTableItemGroup.swift [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+        // Set all the previously selected items as selected
+        guard let selectableItems = self.items as? [RSDChoiceTableItem] else { return }
+        for input in selectableItems {
+            input.selected = input.choice.isEqualToResult(result)
+        }
+    }
+    
+    /// Select or de-select an item (answer) at a specific indexPath. This is used for text choice and boolean answers.
+    /// - parameters:
+    ///     - selected:   A bool indicating if item should be selected
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItemGroup.swift]: .

File: ResearchSuite/ResearchSuite/RSDTableItemGroup.swift [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+    ///     - indexPath:  The IndexPath of the item
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableItemGroup.swift]: .

File: ResearchSuite/ResearchSuite/RSDTableSection.swift [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    open private(set) var itemGroups: [RSDTableItemGroup] = []
+    
+    /// The table section index.
+    open private(set) var index: Int
+    
+    /// The title for this section.
+    public var title: String?
+    
+    /// Returns the total count of all Items in this section.
+    // - returns: The total number of RSDGenericStepTableItems in this section
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTableSection.swift]: .

File: ResearchSuite/ResearchSuite/RSDTaskController.swift [hunk 20]
```diff
@@ -33,123 +33,103 @@
...
-     Optional factory subclass that can be used to vend custom steps that are decoded from a plist or json.
-     */
+    /// Optional factory subclass that can be used to vend custom steps that are decoded
+    /// from a plist or json.
     var factory: RSDFactory? { get }
     
-    /**
-     Returns the currently active step controller (if any)
-     */
+    /// Returns the currently active step controller (if any)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTaskController.swift]: .

File: ResearchSuite/ResearchSuite/RSDTaskInfoStepObject.swift [hunk 21]
```diff
@@ -53,52 +50,90 @@ public struct RSDTaskInfoStepObject : RSDTaskInfoStep, RSDSchemaInfo, Decodable
...
     }
+    private var _schemaIdentifier: String?
+    
+    /// A revision number associated with the result schema.
     public var schemaRevision: Int = 1
     
+    /// The task transformer for vending a task.
     public var taskTransformer: RSDTaskTransformer!
     
+    /// The estimated fetch time is determined by the task transformer
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTaskInfoStepObject.swift]: .

File: ResearchSuite/ResearchSuite/RSDUIStepObject.swift [hunk 22]
```diff
@@ -33,56 +33,170 @@
...
     
+    /// Instantiate a step result that is appropriate for this step. Default implementation will return a `RSDResultObject`.
+    /// - returns: A result for this step.
     open func instantiateStepResult() -> RSDResult {
         return RSDResultObject(identifier: identifier)
     }
     
     // MARK: validation
     
+    /// Required method. The base class implmentation does nothing.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUIStepObject.swift]: impl**e**mentation

File: ResearchSuite/ResearchSuite/RSDUIThemeElement.swift [hunk 23]
```diff
@@ -74,11 +74,21 @@ public protocol RSDColorThemeElement : RSDUIThemeElement {
     
     /// The background color for this step. If undefined then the background color appropriate to the light
     /// style will be used.
+    /// - returns: The color or `nil` if undefined
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUIThemeElement.swift]: .

File: ResearchSuite/ResearchSuite/RSDUIThemeElement.swift [hunk 24]
```diff
@@ -74,11 +74,21 @@ public protocol RSDColorThemeElement : RSDUIThemeElement {
...
+    #if os(watchOS)
+    func backgroundColor() -> UIColor?
+    #else
     func backgroundColor(compatibleWith traitCollection: UITraitCollection?) -> UIColor?
+    #endif
     
     /// The foreground color for this step. If undefined then the foreground color appropriate to the light
     /// style will be used.
+    /// - returns: The color or `nil` if undefined
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDUIThemeElement.swift]: .

File: ResearchSuite/ResearchSuite/RSDViewThemeElementObject.swift [hunk 25]
```diff
@@ -33,15 +33,70 @@
...
+    /// via the `viewIdentifier`.
     public let storyboardIdentifier: String?
     
+    private enum CodingKeys: String, CodingKey {
+        case viewIdentifier, bundleIdentifier, storyboardIdentifier
+    }
+    
+    /// Default initializer.
+    ///
+    /// - paramaters:
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDViewThemeElementObject.swift]: param**e**ters

File: ResearchSuite/ResearchSuite/UIColor+Utilities.swift [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    /// Finally, the method will look for a file imbedded in the given bundle named "ColorInfo.plist".
+    /// This allows the app to define a mapping of colors in a single place using a plist with key/value pairs.
+    /// The mapping is expected to include key/value pairs where the value is a hex-coded string. The plist
+    /// method is slower (because the file must be openned and inspected) so it is only recommended for applications
+    /// that need to be reverse-compatible to iOS 10.
+    ///
+    /// - parameters:
+    ///     - name: The name of the color. Either a HEX-code or a custom-defined key.
+    ///     - bundle: The bundle with either the Color asset (iOS 11) or the "ColorInfo.plist" file (iOS 10)
+    ///     - traitCollection: The trait collection (if supported)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/UIColor+Utilities.swift]: .

File: ResearchSuite/ResearchSuite/UIColor+Utilities.swift [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+    ///
+    /// Finally, the method will look for a file imbedded in the given bundle named "ColorInfo.plist".
+    /// This allows the app to define a mapping of colors in a single place using a plist with key/value pairs.
+    /// The mapping is expected to include key/value pairs where the value is a hex-coded string. The plist
+    /// method is slower (because the file must be openned and inspected) so it is only recommended for applications
+    /// that need to be reverse-compatible to iOS 10.
+    ///
+    /// - parameters:
+    ///     - name: The name of the color. Either a HEX-code or a custom-defined key.
+    ///     - bundle: The bundle with either the Color asset (iOS 11) or the "ColorInfo.plist" file (iOS 10)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/UIColor+Utilities.swift]: .

**Review by Erin-Mounts — APPROVED** [review]
Buncha nits and one question, otherwise looks good, merge when ready

File: ResearchSuite/ResearchSuite/RSDInputField.swift (hunk 3 above)
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuite/RSDInputField.swift:89]: Nope, it is generally implemented in place of a skip. Basically, where you want to have the answer be skippable but include the response in the list of choices. It's a UI thing. This allows for customization of that text.

**Pull request merged**